File: packages/router/src/adapters/openai.ts
export type OpenAIConfig = {
  apiKey?: string
  baseUrl: string
  model: string
}


function getEnv(
  key: string
): string | undefined {

  if (
    typeof globalThis !== "undefined" &&
    "process" in globalThis
  ) {
    const env =
      (globalThis as {
        process?: {
          env?: Record<string, string | undefined>
        }
      }).process?.env

    return env?.[key]
  }

  return undefined
}


export function getOpenAIConfig(): OpenAIConfig {
  return {
    apiKey: getEnv(
      "OPENAI_API_KEY"
    ),

    baseUrl:
      getEnv(
        "OPENAI_BASE_URL"
      ) ||
      "https://api.openai.com/v1",

    model:
      getEnv(
        "OPENAI_MODEL"
      ) ||
      "gpt-4o-mini"
  }
}


export async function openAIExecute(
  message: string
) {
  const config = getOpenAIConfig()

  if (!config.apiKey) {
    return {
      provider: "openai",
      response:
        "OpenAI provider configured but API key missing"
    }
  }


  const response = await fetch(
    `${config.baseUrl}/chat/completions`,
    {
      method: "POST",

      headers: {
        "Content-Type":
          "application/json",

        Authorization:
          `Bearer ${config.apiKey}`
      },

      body: JSON.stringify({
        model: config.model,

        messages: [
          {
            role: "user",
            content: message
          }
        ]
      })
    }
  )


  const data =
    await response.json()


  return {
    provider: "openai",
    response:
      data.choices?.[0]?.message?.content ||
      "No response"
  }
}

File: agents/api/providers/openai.py
import json
import os
import urllib.error
import urllib.request


def health():
    return bool(
        os.getenv("OPENAI_API_KEY")
    )


def run(message: str):

    key = os.getenv(
        "OPENAI_API_KEY"
    )

    if not key:
        return {
            "provider": "openai",
            "success": False,
            "response":
                "OpenAI provider unavailable: missing OPENAI_API_KEY"
        }

    try:
        request = urllib.request.Request(
            "https://api.openai.com/v1/responses",
            data=json.dumps({
                "model": "gpt-5-mini",
                "input": message
            }).encode(),
            headers={
                "Authorization": f"Bearer {key}",
                "Content-Type": "application/json"
            },
            method="POST"
        )

        with urllib.request.urlopen(request, timeout=30) as response:
            data = json.loads(response.read())

        if "output" in data:
            text = data["output"][0]["content"][0]["text"]
        else:
            text = str(data)

        return {
            "provider": "openai",
            "success": True,
            "response": text
        }

    except urllib.error.HTTPError as e:
        return {
            "provider": "openai",
            "success": False,
            "response": f"{e.code} {e.reason}: {e.read().decode(errors='replace')}"
        }

    except Exception as e:
        return {
            "provider": "openai",
            "success": False,
            "response": str(e)
        }
